### ISSUE TITLE
less Google results depending on the SearX language option

### ISSUE THREAD

**Issue by WAZAAAAA0** [searx/searx#1236]
Searching for the query `"Forum cliccando sul tasto rosso"` (with the inverted quotes), which is a random unique Italian sentence from a .it domain that should only find 1-2 results... might or might not find anything on SearX depending on the current language setting.

I searched the same query on every language supported by searx.me 0.14.0 with the default preferences. Only the searches performed with the Italian language settings (it, it-CH, it-IT) and for some reason Chinese (zh, zh-CN, zh-HK, zh-TW) found the result.
Strangely enough, that query seems to work correctly with Google Images in every language.

Next, I also tried these two unique queries in different languages:
`"Или официальное название превыше всего?"`
`(from http://ingvarr.net.ru/forum/7-251-1)`
`"اقصد يعني اذا انت عاوز تسوق مشروعك لازم ان تزيل هذه الكلمات لكي يكون البرنامج احترافي"`
`(from http://arabteam2000-forum.com/index.php?/topic/265383-%D9%83%D9%8A%D9%81-%D8%A3%D9%82%D9%88%D9%85-%D8%A8%D8%AA%D8%B3%D9%88%D9%8A%D9%82-%D8%A8%D8%B1%D9%86%D8%A7%D9%85%D8%AC-%D8%A7%D9%84%D9%85%D8%AD%D8%A7%D8%B3%D8%A8%D8%A9-%D8%A7%D9%84%D8%AE%D8%A7%D8%B5-%D8%A8%D9%8A%D8%9F/)`

The Russian query could only be found with the Russian setting and Chinese. The Arabic one only in Arabic... and Chinese!
It should be noted that official google.com/google.it were able to find all 3 of the example queries I provided in different languages as expected.

I noticed that searx/searx/engines/google.py line 80 (google.cn reference) is commented out. Might this somehow be related to what is going on? Can we just comment out every country domain and let SearX use the default google.com for every language instead? Because Chinese-Google-SearX would be the optimal way to search, if only it didn't give... well... priority to results in a different language than your own.

TL;DR
Google through SearX in Chinese gets more results, can we fix every other language?

**Comment by WAZAAAAA0** [comment]
also, the **Number of results:** section starts working normally in Chinese

**Comment by dalf** [comment]
Thank you for your report !

I've some difficulties to reproduce the examples you give. 
Did you used the [Image category](https://searx.me/?q=%22%D0%98%D0%BB%D0%B8+%D0%BE%D1%84%D0%B8%D1%86%D0%B8%D0%B0%D0%BB%D1%8C%D0%BD%D0%BE%D0%B5+%D0%BD%D0%B0%D0%B7%D0%B2%D0%B0%D0%BD%D0%B8%D0%B5+%D0%BF%D1%80%D0%B5%D0%B2%D1%8B%D1%88%D0%B5+%D0%B2%D1%81%D0%B5%D0%B3%D0%BE%3F%22&time_range=&language=en-US&category_images=on&theme=oscar) or the [General category](https://searx.me/?q=%22%D0%98%D0%BB%D0%B8+%D0%BE%D1%84%D0%B8%D1%86%D0%B8%D0%B0%D0%BB%D1%8C%D0%BD%D0%BE%D0%B5+%D0%BD%D0%B0%D0%B7%D0%B2%D0%B0%D0%BD%D0%B8%D0%B5+%D0%BF%D1%80%D0%B5%D0%B2%D1%8B%D1%88%D0%B5+%D0%B2%D1%81%D0%B5%D0%B3%D0%BE%3F%22&language=en-US&time_range=&category_general=on&safesearch=0&theme=oscar) ? In both cases, I have some results in English.

Which instance have you used ? or better, can you provide some links?

About google.cn: it just shows a redirect page to www.google.com.hk (the Hong Kong version).  Google is not allowed in mainland China.

**Comment by WAZAAAAA0** [comment]
Alright I should've used search links in my report, here they are:

[en-US search:](https://searx.me/?q=%22%D0%98%D0%BB%D0%B8%20%D0%BE%D1%84%D0%B8%D1%86%D0%B8%D0%B0%D0%BB%D1%8C%D0%BD%D0%BE%D0%B5%20%D0%BD%D0%B0%D0%B7%D0%B2%D0%B0%D0%BD%D0%B8%D0%B5%20%D0%BF%D1%80%D0%B5%D0%B2%D1%8B%D1%88%D0%B5%20%D0%B2%D1%81%D0%B5%D0%B3%D0%BE%3F%22&categories=general&language=en-US) only finds unrelated trash, most probably coming from results without quotes
[zh search:](https://searx.me/?q=%22%D0%98%D0%BB%D0%B8%20%D0%BE%D1%84%D0%B8%D1%86%D0%B8%D0%B0%D0%BB%D1%8C%D0%BD%D0%BE%D0%B5%20%D0%BD%D0%B0%D0%B7%D0%B2%D0%B0%D0%BD%D0%B8%D0%B5%20%D0%BF%D1%80%D0%B5%D0%B2%D1%8B%D1%88%D0%B5%20%D0%B2%D1%81%D0%B5%D0%B3%D0%BE%3F%22&categories=general&language=zh) found that 1 unique result word by word! No garbage results. "Number of results" counter also working correctly

Have a side-by-side comparison too:
![enVSzh](https://i.imgur.com/Q52arfJ.png)

So which Google does SearX use when set to Chinese then? I assumed it falls back to google.com from that file. If this is true, then google.com works so much better than any other country-specific google.xx domain with SearX! Maybe SearX is using country domains erroneously by enabling Google's `Pages from: X-COUNTRY` setting? That's my best bet.

**Comment by Pofilo** [comment]
Take a look at:
[https://github.com/asciimoo/searx/blob/master/searx/engines/google.py](https://github.com/asciimoo/searx/blob/master/searx/engines/google.py)

As `CN` is not used, I understand it takes the default one, but the result is not coherent.

**Comment by WAZAAAAA0** [comment]
Today the Russian query temporarily stopped finding the correct result for some reason, so I made more example links with the Italian one. "Chinese" being more comprehensive is still real:

QUERY: `"Forum cliccando sul tasto rosso"`
SOURCE: `http://orologi.forumfree.it/?t=72189596`
[中文 - zh](https://searx.me/?q=%22Forum%20cliccando%20sul%20tasto%20rosso%22&categories=general&language=zh)
[English (United States) - en-US](https://searx.me/?q=%22Forum%20cliccando%20sul%20tasto%20rosso%22&categories=general&language=en-US)
EDIT: it seems like en-US now finds it for some reason, but the problem still exists:
[Español - es](https://searx.me/?q=%22Forum%20cliccando%20sul%20tasto%20rosso%22&categories=general&language=es)
[Français - fr](https://searx.me/?q=%22Forum%20cliccando%20sul%20tasto%20rosso%22&categories=general&language=fr)

I really believe this is a major flaw in SearX that should be taken care of. I've been using SearX in Chinese since I found about this problem. I've seen this problem occur to other metasearch engines too like StartPage and Ixquick, they say they're powered by Google but give less results than the real one, so it's hard to move away from it.

**Comment by WAZAAAAA0** [comment]
I did some further testing and the Chinese language apparently lost its "superpowers" recently on searx.me.
searx.ch which is stuck on `0.11.0` still retains them.

https://searx.me/?q=%21google%20%22Forum%20cliccando%20sul%20tasto%20rosso%22&categories=none&language=zh
https://searx.ch/?q=%21google%20%22Forum%20cliccando%20sul%20tasto%20rosso%22&categories=none&language=zh
![meVSch](https://i.imgur.com/8QbBEiE.png)

I'm gonna be an asshole and tag a bunch of recent contributors to notice this issue because I think it's so important @asciimoo @kvch @MarcAbonce @dalf @Pofilo
I really want to replace Google lol

**Comment by MarcAbonce** [comment]
As you already mentioned, all your queries work if you actually specify the correct language in searx. That is, if you explicitly request all English sites with "Forum cliccando sul tasto rosso", then it will return the English sites with that string. If you request Italian content, the original source in Italian will be shown.

The problem with letting Google "magically" guess the language on our behalf is that it leads to weird behaviour, as you can read on #1085 . For example, search.ch usually returns a mix of German and English results if you let it guess on your behalf, because of the server's location.

So, if we search in google.com, it will ignore the user's chosen language and still return results based either on the server's location or the query. On the other hand, if we enforce a domain, results will always be in the correct language but they will be filtered and may therefore miss more relevant results. Fixing this breaks #1085 and vice versa.

The other alternative would be to add the "Automatic/All" option again, although that has its own set of problems.

**Comment by WAZAAAAA0** [comment]
Thanks @MarcAbonce now that makes a bit more sense.
But I'd still take the magic IP guess over having less results any day personally. I didn't even notice the German IP "bias" while I was using searx in Chinese until you told me... what I did notice was a lot of "annoying" Chinese results because some search engines do listen to the setting, and even Google does to an extent.

So if the "Automatic/All" setting still existed I would use that in a heartbeat. What are these other problems related to re-adding this? Just don't make it the default option, maybe it would help?

**Comment by WAZAAAAA0** [comment]
I mean look at these results, Google DOES listen to my Chinese preference (unfortunately lol) even though the server IP location is Switzerland

![ChinaBias](https://i.imgur.com/IdRg2z0.png)

**Comment by unixfox** [comment]
@WAZAAAAA0 is this issue still occurring with the new Google engine: https://github.com/asciimoo/searx/pull/1628?

**Comment by unixfox** [comment]
@return42 stale issue, I think this can be closed because we haven't heard the author about the new code for Google.

**Comment by WAZAAAAA0** [comment]
yo mama stale

I did more testing locally with the latest master version using the Russian search query.
http://127.0.0.1:8888/0q=%21google%20%22%D0%98%D0%BB%D0%B8%20%D0%BE%D1%84%D0%B8%D1%86%D0%B8%D0%B0%D0%BB%D1%8C%D0%BD%D0%BE%D0%B5%20%D0%BD%D0%B0%D0%B7%D0%B2%D0%B0%D0%BD%D0%B8%D0%B5%20%D0%BF%D1%80%D0%B5%D0%B2%D1%8B%D1%88%D0%B5%20%D0%B2%D1%81%D0%B5%D0%B3%D0%BE%3F%22&categories=none&language=ru-RU
On normal google.com, it should find 2 results: a Russian forum post, and this English github issue. 
This is how SearX behaves on each language option:
```
RESULTS - LANGUAGE
2	Default language (my browser is set to the en-US locale)
2	ms-MY
0	ca-ES
0	da-DK
0	de
0	de-AT
0	de-CH
0	de-DE
0	et-EE
1	en
1	en-AU
1	en-CA
1	en-GB
1	en-IN
1	en-MY
2	en-US
0	es
0	es-AR
0	es-ES
0	es-MX
0	fr
0	fr-BE
0	fr-CA
0	fr-CH
0	fr-FR
0	hr-HR
0	id-ID
0	it-IT
0	lv-LV
0	lt-LT
0	hu-HU
0	nl
0	nl-BE
0	nl-NL
0	nb-NO
0	pl-PL
0	pt
0	pt-BR
0	pt-PT
0	ro-RO
0	sk-SK
0	sl-SI
0	fi-FI
0	sv-SE
0	vi-VN
0	tr-TR
0	is-IS
0	cs-CZ
0	el-GR
0	bg-BG
1	ru-RU
0	sr-RS
0	uk-UA
0	he-IL
0	ar-SA
0	fa-IR
0	th-TH
0	zh
0	zh-CN
0	zh-TW
0	ja-JP
0	ko-KR
```
(0 actually means it found trash unrelated results, presumably the `No results found for "A B C". Results for A B C (without quotes):` is coming into play)
Conclusions:
- Russian option only finds the Russian result
- English options (except en-US) only find the English result
- en-US and ms-MY find both
- the rest of the languages only find garbage results

The SearX language option affecting the number of results is still happening, except this time the "Chinese powers" have been transferred to en-US and ms-MY it seems.

**Comment by return42** [comment]
@WAZAAAAA0 thanks for your report. I can't reproduce your experience. With de-DE I got hundreds of hits (from .de server) .. I have read through the PR comments .. but Iam a bit confused .. I can't see a solution nor a real issue.  This might be my fault / feel free to reopen if so. Sending a PR would also be helpful.

**Comment by WAZAAAAA0** [comment]
>I can't reproduce your experience. With de-DE I got hundreds of hits

@return42 But you just reproduced it. Please read those hundreds of hits: they're all garbage. You were supposed to find exactly 2 results (now 3). Only en-US and ms-MY do that, try to compare them to a normal google.com search, or just look at this picture:
![LanguageDifferences](https://i.imgur.com/InPAEcs.png)

Issue is not yet solved, but using en-US (or ms-MY) is a good workaround for now. I feel sorry for anyone who uses any other language.

**Comment by WAZAAAAA0** [comment]
Still happening as of today, tested on various instances.

1 result on en-US:
https://searx.sunless.cloud/?q=%21google%20galgoutta&categories=none&language=en-US
0 results on other languages:
https://searx.sunless.cloud/?q=%21google%20galgoutta&categories=none&language=es-ES

EDIT: Google itself now shows 0 results, this query is no longer a valid test
EDIT2: now it's valid again

**Comment by return42** [comment]
we are working on the google engine  .. see #1985 / you can test it here: https://darmarit.org/searx/?q=%21go%20galgoutta&categories=none&language=en-US

**Comment by unixfox** [comment]
@WAZAAAAA0 could you test again now that #1985 got merged?

**Comment by WAZAAAAA0** [comment]
@return42 @unixfox Tested the Russian sentence query at the darmarit.org/searx instance on 2020-08-14 with version 0.17.0 

Searching that on the real google.com currently finds 2 results; one link to this GitHub issue, and one link to the Russian forum. Here's the outcome:
```
2 results (GitHub issue + Russian forum):
am-ET az-AZ bn-BD br-FR bs-BA cy-GB eu-ES fil-PH fo-FO gd-GB ka-GE kk-KZ km-KH ky-KG lo-LA mk-MK mn-MN ms-MY mt-MT ne-NP pt pt-BR pt-PT rw-RW si-LK so-SO sq-AL sr-ME tg-TJ ti-ER tk-TM ur-PK uz-UZ

1 result (GitHub issue):
en en-AU en-CA en-GB en-IE en-IN en-NZ en-PH en-SG en-US

1 result (Russian forum):
ru-RU

0 results (garbage/unrelated): 
af-NA ar-SA be-BY bg-BG ca-AD cs-CZ da-DK de de-AT de-CH de-DE el-GR es es-AR es-CL es-ES es-MX et-EE fa-IR fi-FI fr fr-BE fr-CA fr-CH fr-FR he-IL hr-HR hu-HU hy-AM id-ID is-IS it-IT ja-JP ko-KR lt-LT lv-LV nb-NO nl nl-BE nl-NL pl-PL ro-RO sk-SK sl-SI sv-SE sw-KE th-TH tr-TR uk-UA vi-VN zh zh-CN zh-TW
```
(clickable examples: [0 results](https://darmarit.org/searx/?q=%21go%20%22%D0%98%D0%BB%D0%B8%20%D0%BE%D1%84%D0%B8%D1%86%D0%B8%D0%B0%D0%BB%D1%8C%D0%BD%D0%BE%D0%B5%20%D0%BD%D0%B0%D0%B7%D0%B2%D0%B0%D0%BD%D0%B8%D0%B5%20%D0%BF%D1%80%D0%B5%D0%B2%D1%8B%D1%88%D0%B5%20%D0%B2%D1%81%D0%B5%D0%B3%D0%BE%3F%22&categories=none&language=sl-SI) VS [2 results](https://darmarit.org/searx/?q=%21go%20%22%D0%98%D0%BB%D0%B8%20%D0%BE%D1%84%D0%B8%D1%86%D0%B8%D0%B0%D0%BB%D1%8C%D0%BD%D0%BE%D0%B5%20%D0%BD%D0%B0%D0%B7%D0%B2%D0%B0%D0%BD%D0%B8%D0%B5%20%D0%BF%D1%80%D0%B5%D0%B2%D1%8B%D1%88%D0%B5%20%D0%B2%D1%81%D0%B5%D0%B3%D0%BE%3F%22&categories=none&language=mt-MT))

The issue of varying number of results depending on language still exists, the currently "blessed" languages have simply shifted.

**Comment by WAZAAAAA0** [comment]
Finds out the folks at the Whoogle project (<https://github.com/benbusby/whoogle-search>) have sort of documented the odd behavior I demonstrated in this issue.

One of the options explain:
```
Filter Results by Country:
Default (none)
— Note: If enabled, a website will only appear in the results if it is *hosted* in the selected country.
```
This is what's going on in SearX, the language settings are limiting the Google results. Honestly SearX should aim to replicate that `Default (none)` behavior!

**Comment by unixfox** [comment]
> Finds out the folks at the Whoogle project (<https://github.com/benbusby/whoogle-search>) have sort of documented the odd behavior I demonstrated in this issue.
> 
> One of the options explain:
> ```
> Filter Results by Country:
> Default (none)
> — Note: If enabled, a website will only appear in the results if it is *hosted* in the selected country.
> ```
> This is what's going on in SearX, the language settings are limiting the Google results. Honestly SearX should aim to replicate that `Default (none)` behavior!

Do you mean we should introduce a new language called "default" which enable this "default (none)" behavior?

**Comment by WAZAAAAA0** [comment]
> Do you mean we should introduce a new language called "default" which enable this "default (none)" behavior?

Yes exactly, the more results the merrier.

**Comment by unixfox** [comment]
After testing a bit with whoogle, you are right the "default (none)" setting is much better than having to select a language.
More over this doesn't require to select a language because Google will automatically recognize the language based on the entered words and display the results in the correct language.
So much more convenient than having to force a language in the Searx settings every time you make searches in multiple languages.

I'll look into the ability to add this kind of feature on Searx.

**Comment by WAZAAAAA0** [comment]
Fixed with https://github.com/searx/searx/pull/2882 and https://github.com/searxng/searxng/pull/134 🎉

Example of properly working instance that correctly finds an Arabic query from an Arabic site:
https://searx.be/search?q=%21go%20%22%D8%A7%D9%82%D8%B5%D8%AF%20%D9%8A%D8%B9%D9%86%D9%8A%20%D8%A7%D8%B0%D8%A7%20%D8%A7%D9%86%D8%AA%20%D8%B9%D8%A7%D9%88%D8%B2%20%D8%AA%D8%B3%D9%88%D9%82%20%D9%85%D8%B4%D8%B1%D9%88%D8%B9%D9%83%20%D9%84%D8%A7%D8%B2%D9%85%20%D8%A7%D9%86%20%D8%AA%D8%B2%D9%8A%D9%84%20%D9%87%D8%B0%D9%87%20%D8%A7%D9%84%D9%83%D9%84%D9%85%D8%A7%D8%AA%20%D9%84%D9%83%D9%8A%20%D9%8A%D9%83%D9%88%D9%86%20%D8%A7%D9%84%D8%A8%D8%B1%D9%86%D8%A7%D9%85%D8%AC%20%D8%A7%D8%AD%D8%AA%D8%B1%D8%A7%D9%81%D9%8A%22&categories=none&language=def

To use the fix, make sure you're using a SearX instance that's enough up-to-date such as `1.0.0-192-c0ed217d` or a `SearXNG` one, and set your `Search language` to `Default language`.
